packages/graphql-codegen-cli/src/utils/patterns.ts

```diff
@@ -0,0 +1,308 @@
+ *    a match, after considering the precedence of any global and local negations
+ */
+export const makeShouldRebuild = ({
+  globalPatternSet,
+  localPatternSets,
+}: {
+  globalPatternSet: PatternSet;
+  localPatternSets: PatternSet[];
+}) => {
+  const localMatchers = localPatternSets.map(localPatternSet => {
+    return (path: string) => {
+      // Is path negated by any negating watch pattern?
+      if (
+        matchesAnyAffirmativePattern(
+          path,
+          invertNegatedPatterns([...globalPatternSet.watch.negated, ...localPatternSet.watch.negated])
+        )
+      ) {
+        // Short circut: negations in watch patterns take priority
+        return false;
+      }
+
+      // Does path match any affirmative watch pattern?
+      if (
+        matchesAnyAffirmativePattern(path, [
+          ...globalPatternSet.watch.affirmative,
+          ...localPatternSet.watch.affirmative,
+        ])
+      ) {
+        // Immediately return true: Watch pattern takes priority, even if documents or schema would negate it
+        return true;
+      }
+
+      // Does path match documents patterns (without being negated)?
+      if (pathMatches(path, [...globalPatternSet.documents.patterns, ...localPatternSet.documents.patterns])) {
+        return true;
+      }
+
+      // Does path match schemas patterns (without being negated)?
+      if (pathMatches(path, [...globalPatternSet.schemas.patterns, ...localPatternSet.schemas.patterns])) {
+        return true;
+      }
+
+      // Otherwise, there is no match
+      return false;
+    };
+  });
+
+  /**
+   * Return `true` if `path` should trigger a rebuild
+   */
+  return ({ path: absolutePath }: { path: string }) => {
+    if (!isAbsolute(absolutePath)) {
+      throw new Error('shouldRebuild trigger should be called with absolute path');
+    }
+
+    const path = relative(process.cwd(), absolutePath);
+    const shouldRebuild = localMatchers.some(matcher => matcher(path));
+    return shouldRebuild;
+  };
+};
+
+/**
+ * Create the pattern set for the "global" (top level) config.
+ *
+ * In the `shouldRebuild` algorithm, any of these watch patterns will take
+ * precedence over local configs, and any schemas and documents patterns will be
+ * mixed into the pattern set of each local config.
+ */
+export const makeGlobalPatternSet = (initialContext: CodegenContext) => {
+  const config: Types.Config & { configFilePath?: string } = initialContext.getConfig();
+
+  return {
+    watch: sortPatterns([
+      ...(typeof config.watch === 'boolean' ? [] : normalizeInstanceOrArray<string>(config.watch ?? [])),
+      relative(process.cwd(), initialContext.filepath),
+    ]),
+    schemas: sortPatterns(makePatternsFromSchemas(normalizeInstanceOrArray<Types.Schema>(config.schema))),
+    documents: sortPatterns(
+      makePatternsFromDocuments(normalizeInstanceOrArray<Types.OperationDocument>(config.documents))
+    ),
+  };
+};
```

Comment:
We set the global "watch" patterns to any defined in `global.watch` (if it's a non-boolean), and also adds the path to the config file. Note that a relative path is used for the config file because we expect all patterns to use relative paths.

Note that we are "sorting" these patterns (grouping them into `patterns`/`affirmative`/`negated` only once, on initialization, so that we can avoid doing that during every path change when we check if a rebuild should be triggered.

Reply:
See https://github.com/dotansimha/graphql-code-generator/pull/9275#discussion_r1181497873